app/src/main/kotlin/compiler/semantic_analysis/ArgumentResolver.kt

```diff
@@ -2,17 +2,182 @@ package compiler.semantic_analysis
+                foundNamedArgument = true
+            } else if (foundNamedArgument) {
+                diagnostics.report(Diagnostic.ArgumentResolutionError.PositionalArgumentAfterNamed(functionCall))
+                return
+            }
+        }
+
+        val function: Function = nameResolution[functionCall] as Function
+        val parameterNames = function.parameters.map { it.name }
+        val numberOfParameters = parameterNames.size
+        val isMatched = parameterNames.map { false }.toMutableList()
+
+        for ((index, argument) in functionCall.arguments.withIndex()) {
+            if (argument.name == null) {
+                if (index >= numberOfParameters) {
+                    diagnostics.report(Diagnostic.ArgumentResolutionError.TooManyArguments(functionCall))
+                    return
+                }
+                result[argument] = function.parameters[index]
+                isMatched[index] = true
+            } else {
+                val parameterIndex = parameterNames.indexOf(argument.name)
+                if (parameterIndex == -1) {
+                    diagnostics.report(Diagnostic.ArgumentResolutionError.UnknownArgument(functionCall, argument.name))
+                    return
+                }
+                if (isMatched[parameterIndex]) {
+                    diagnostics.report(Diagnostic.ArgumentResolutionError.RepeatedArgument(functionCall, argument.name))
+                    return
+                }
+                result[argument] = function.parameters[parameterIndex]
+                isMatched[parameterIndex] = true
+            }
+        }
+
+        for (index in 0 until numberOfParameters) {
+            if (!isMatched[index] && function.parameters[index].defaultValue == null) {
+                diagnostics.report(Diagnostic.ArgumentResolutionError.MissingArgument(functionCall, parameterNames[index]))
+            }
+
+            // TODO: detect default parameter values used in this function call
+        }
+    }
+
+    private fun checkExpression(expression: Expression) {
+        when (expression) {
+            is Expression.FunctionCall -> {
+                checkFunctionCall(expression)
+            }
+            is Expression.UnaryOperation -> {
+                checkExpression(expression.operand)
+            }
+            is Expression.BinaryOperation -> {
+                checkExpression(expression.leftOperand)
+                checkExpression(expression.rightOperand)
+            }
+            is Expression.Conditional -> {
+                checkExpression(expression.condition)
+                checkExpression(expression.resultWhenTrue)
+                checkExpression(expression.resultWhenFalse)
+            }
+        }
+    }
+
+    private fun checkVariable(variable: Variable) {
+        if (variable.value != null) {
+            checkExpression(variable.value)
+        }
+    }
+
+    private fun checkFunctionCalls(program: Program) {
```

Comment:
I'd say the main responsibility of this function is to map arguments to parameters so how about a name `resolveFunctionCallsArguments`?